README.md

```diff
@@ -220,3 +220,433 @@ If a page number is out of bounds, an empty list is returned.
+is mainly used to update the summary cache and resource counts, but any module
+could add a handler to react to any resource change or get cached cluster data.
+For example, if we wanted a handler to log all "add" events for newly created
+secrets:
+
+```go
+import (
+	"context"
+	"github.com/rancher/steve/pkg/server"
+	"k8s.io/apimachinery/pkg/runtime"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+func logSecretEvents(server *server.Server) {
+	server.ClusterCache.OnAdd(context.TODO(), func(gvk schema.GroupVersionKind, key string, obj runtime.Object) error {
+		if gvk.Kind == "Secret" {
+			logrus.Infof("[event] add: %s", key)
+		}
+		return nil
+	})
+}
+```
+
+### Aggregation
+
+Rancher uses a concept called "aggregation" to maintain connections to remote
+services. Steve implements an aggregation client in order to allow connections
+from Rancher and expose its API to Rancher.
+
+Aggregation is enabled by defining a secret name and namespace in the steve
+server:
+
+```go
+import (
+	"context"
+	"github.com/rancher/steve/pkg/server"
+)
+
+func run() {
+	restConfig := getRestConfig()
+	server := server.New(context.TODO(), restConfig, &server.Options{
+		AggregationSecretNamespace: "cattle-system",
+		AggregationSecretName: "stv-aggregation",
+	})
+	server.ListenAndServe(context.TODO(), 9443, 9080, nil)
+}
+```
+
+This prompts the steve server to start a controller that watches for this
+secret. The secret is expected to contain two pieces of data, a URL and a
+token:
+
+```sh
+$ kubectl -n cattle-system get secret stv-aggregation -o yaml
+apiVersion: v1
+data:
+  token: Zm9vYmFy
+  url: aHR0cHM6Ly8xNzIuMTcuMC4xOjg0NDMvdjMvY29ubmVjdA==
+kind: Secret
+metadata:
+...
+```
+
+Steve makes a websocket connection to the URL using the token to authenticate.
+When the secret changes, the steve aggregation server restarts with the
+up-to-date URL and token.
+
+Through this websocket connection, the steve agent is exposed on the remote
+management server and the management server can route steve requests to it. The
+management server can also keep track of the availability of the agent by
+detecting whether the websocket session is still active. In Rancher, the
+connection endpoint runs on /v3/connect.
+
+Rancher implements aggregation for other types of services as well. In Rancher,
+the user can define endpoints via a
+[v3.APIService](https://pkg.go.dev/github.com/rancher/rancher/pkg/apis/management.cattle.io/v3#APIService)
+custom resource (which is distinct from the built-in Kubernetes
+[v1.APIService](https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/api-service-v1/)
+resource. Then Rancher runs a middleware handler that routes incoming requests
+to defined endpoints. The external services follow the same process of using a
+defined secret containing a URL and token to connect and authenticate to
+Rancher. This aggregation is defined independently and does not use steve's
+aggregation client.
```

Comment:
Nit: feels like this portion could just be in the rancher README.md.

Reply:
We could add it there as well, but I and others have been confused about "steve aggregation" and what it really means, since there is a significant amount of code in steve dedicated to this not-steve functionality, so I prefer to keep it here.